analyzer/evmverifier/sourcify/client.go

```diff
@@ -0,0 +1,210 @@
+	common.ChainNameTestnet: {
+		common.RuntimeEmerald:  "42261",
+		common.RuntimeSapphire: "23295",
+	},
+	common.ChainNameMainnet: {
+		common.RuntimeEmerald:  "42262",
+		common.RuntimeSapphire: "23294",
+	},
+}
+
+// SourcifyClient is a client for interacting with the Sourcify Server API,
+// providing methods to fetch and parse contract data for supported Oasis runtimes.
+type SourcifyClient struct {
+	serverUrl *url.URL
+
+	chain common.ChainName
+
+	httpClient *http.Client
+
+	logger *log.Logger
+}
+
+func (s *SourcifyClient) callAPI(ctx context.Context, method string, url *url.URL) ([]byte, error) {
+	s.logger.Debug("sourcify API call", "method", method, "url", url.String())
+	req, err := http.NewRequestWithContext(ctx, method, url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	res, err := s.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("sourcify API call failure: %w", err)
+	}
+	defer res.Body.Close()
+	resp, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	s.logger.Debug("sourcify API call response", "method", method, "url", url.String(), "status", res.Status, "response_body", string(resp))
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("sourcify API call failure: %s", res.Status)
+	}
+
+	return resp, nil
+}
+
+// GetVerifiedContractAddresses returns a list of all verified contract addresses for the given runtime.
+//
+// API docs: https://docs.sourcify.dev/docs/api/server/get-contract-addresses-all/
+//
+// Note: This uses the free, public server API. If it turns out to be unreliable, we could use the repository API (vis IPFS proxy) instead, e.g.:
+// http://ipfs.default:8080/ipns/repo.sourcify.dev/contracts/full_match/23294
+func (s *SourcifyClient) GetVerifiedContractAddresses(ctx context.Context, runtime common.Runtime) ([]ethCommon.Address, error) {
+	// Fetch verified contract addresses.
+	u := *s.serverUrl
+	u.Path = path.Join(u.Path, "files/contracts", sourcifyChains[s.chain][runtime])
+	body, err := s.callAPI(ctx, http.MethodGet, &u)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch verified contract addresses: %w", err)
+	}
+
+	// Parse response.
+	var response struct {
+		Full []ethCommon.Address `json:"full"`
+		// XXX: Skip partial matches for now: https://docs.sourcify.dev/docs/full-vs-partial-match/
+	}
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nil, fmt.Errorf("failed to parse verified contract addresses: %w", err)
+	}
+
+	return response.Full, nil
+}
+
+// Metadata represents a subset of the metadata.json file generated by the Solidity compiler.
+// The exact structure of the metadata output can vary depending on the Solidity compiler version,
+// and there is no standard format. However, the fields included in this struct should generally present
+// across different compiler versions and contracts.
+type Metadata struct {
```

Comment:
IDK how deep we want to go with the parsing here.

FWIW, Etherscan ([example](https://etherscan.io/address/0xa0b319a07521c43c4961a1a76977a5ccd46c3122#code)) currently shows the following fields:
- contract name (:question: ; not sure where that comes from, maybe they parse the `.sol`. We don't need to do that.)
- compiler version :heavy_check_mark:
- optimizations enabled :heavy_check_mark: (we have this inside `settings.optimizer`)
- settings :heavy_check_mark: 

So if we wanted to match Etherscan, the current struct provides an unusual mix of parsedness: Some fields are explicitly parsed (compiler version), some are available inside an untyped JSON but with an expected key (optimizer), and some are completely untyped (settings). (Edit: I see elsewhere that you might be thinking of a more structured representation in the DB and/or the API.)

Currently I'm leaning towards the opinion that it's fine (and even desirable) to return the entire metadata as a blob, and let the clients do the parsing:
- From the indexer perspective: Beyond the ABI, the indexer doesn't care about the contents of metadata.json; it won't allow querying by individual fields, we won't compute aggregate stats or similar ... we'll just be relaying (a part of) the metadata, unchanged.
- From the client perspective: They need to retrieve the information from the indexer's JSON. They won't care much whether they need to follow our openapi for that, or Solidity's metadata API, or a combination of both. But I'm thinking if we give them the entire metadata.json as-is, they might be able to use existing eth libraries to inspect it.
- The information that is relevant for our Explorer is still easy to access, but information that might be of interest to other clients is also still there.
- As a bonus, it's the easiest to implement and maintain on the indexer side.

Reply:
Agree. We'll likely never care about the contents on the indexer so makes sense to store the whole JSON. 

So all in all, I'm now thinking of having the following fields in the DB:

```
abi JSONB
metadata_json JSONB
source_files BYTEA
```

Reply:
source_files will also be JSONB